Button to display message source in the New Modmail

Xeoth opened the pull request:
If I did everything right, this fixes #323

creesch commented on extension/data/modules/newmodmailpro.js:
```diff
@@ -390,6 +390,40 @@ function newmodmailpro () {
+
+                // Fetch and store the conversation info in cache
+                const currentID = event.detail.pageDetails.conversationID;
+                TBStorage.setCache('NewModmailPro', 'current-conversation', await TBApi.apiOauthGET(`/api/mod/conversations/${currentID}`).then(r => r.json()));
```
I am not sure we need to do the request for every modmail loaded. Possibly just do the request the first time source is clicked

creesch replied:
You haven't actually resolved this, for every message in a conversation you are still doing a request. 

My suggestion would be to set up your cache something along these lines: 

1. Keep a local boolean named `conversationCached` and set it to `false` by default.
2. Attach the buttons to the replies.
3. When a source button is clicked check the boolean, if true use cache for the source if false fetch the conversation and set the cache. 
4. Clear the boolean based on a `TBNewPage`  event.

creesch commented on extension/data/modules/newmodmailpro.js:
```diff
@@ -390,6 +390,40 @@ function newmodmailpro () {
+        window.addEventListener('TBNewPage', async event => {
+            if (event.detail.pageType === 'modmailConversation') {
+                // Add the button to the buttons over the message
+                $('.Thread__message').append('<button class="tb-source-button tb-action-button">Show Source</button>');
```
You also have conversations with collapsed messages for which you need to account. We already do this elsewhere in toolbox, I'll look it up in a bit.

Xeoth commented:
Thanks for your comments, Creesch. I wouldn't have noticed those. I fixed what needed fixing and added a toggling feature. It should all be good now.

Please review and leave a comment if you find something that still needs a fix.

creesch commented on extension/data/modules/newmodmailpro.js:
```diff
@@ -390,6 +397,52 @@ function newmodmailpro () {
+                const currentID = event.detail.data.post.id;
+                TBStorage.setCache('NewModmailPro', 'current-conversation', await TBApi.apiOauthGET(`/api/mod/conversations/${currentID}`).then(r => r.json()));
+
+                $('.tb-source-button').unbind().click(async e => {
```
`unbind` is actually deprecated. You probably found it googling as the listener is triggered multiple times. A different solution would be to use `e.stopImmediatePropagation()`

creesch commented on extension/data/modules/newmodmailpro.js:
```diff
@@ -390,6 +397,52 @@ function newmodmailpro () {
+                    const $currentSourceBtn = $(e.currentTarget.parentElement);
+
+                    // Getting the ID of the message on which the button was clicked
+                    const [activeMessageID] = $currentSourceBtn.find('.Message__date')[0].pathname.split('/').slice(-1);
```
You already have the id from the jsAPI event, so this isn't really needed.

creesch commented on extension/data/modules/newmodmailpro.js:
```diff
@@ -390,6 +397,52 @@ function newmodmailpro () {
+
+                const $target = $(event.target);
+
+                $target.closest('.Thread__message').append('<button class="tb-source-button tb-action-button">Toggle Source</button>');
```
I really don't like the placement of this button, my suggestion would be to add it to the message header and change the class to `tb-general-button`. 

Like so 

![image](https://user-images.githubusercontent.com/2502190/90731185-a3a49e80-e2c9-11ea-91bf-cf6f6e5a74e3.png)

creesch commented on extension/data/styles/newmodmailpro.css:
```diff
@@ -36,6 +36,22 @@
+/* source textarea
+*/
+
+.tb-source-button {
```
Following my other comment this margin needs to change so it fits in the header.

The quote and Report button both have a left margin of 16px 

```css
body.mod-toolbox-rd .tb-source-button {
    margin-left: 16px;
}
```

creesch reviewed (changes requested):
So uh yeah I ended up testing my suggested changes to make sure that what I was saying makes sense. 

So if you want you can simply use the below code ;) 

```javascript
        if (sourceButton) {
            let conversationCached = false;
            window.addEventListener('TBNewPage', () => {
                conversationCached = false;
            });

            TB.listener.on('author', event => {
                if (event.detail.type !== 'TBmodmailCommentAuthor') {
                    return;
                }
                const $target = $(event.target);

                $target.closest('.Message__header').append('<button class="tb-source-button tb-general-button">source</button>');

                $('.tb-source-button').click(async e => {
                    // Something is causing the listener to be triggered multiple times.
                    e.stopImmediatePropagation();
                    const $currentSourceBtn = $(e.currentTarget.parentElement);

                    // Getting the ID of the message on which the button was clicked.
                    const activeMessageID = event.detail.data.comment.id;
                    const $currentSourceField = $(`#tb-source-${activeMessageID}`);

                    // Toggling the source
                    if ($currentSourceField.length) {
                        // If the source field exists, toggle it

                        $currentSourceField.toggle();
                    } else {
                        // If the source field is not present (has not been requested yet), request it and create
                        // a div+textarea with the source.

                        if (!$currentSourceBtn.closest('.Thread__message').has('.tb-source-field').length) {
                            let conversationInfo;
                            if (conversationCached) {
                                conversationInfo = await TBStorage.getCache('NewModmailPro', 'current-conversation');
                            } else {
                                // Fetch and store the conversation info in cache
                                const currentID = event.detail.data.post.id;
                                conversationInfo = await TBApi.apiOauthGET(`/api/mod/conversations/${currentID}`).then(r => r.json());
                                TBStorage.setCache('NewModmailPro', 'current-conversation', conversationInfo);
                                conversationCached = true;
                            }
                            // Getting the body in markdown from selected message
                            const messageSource = conversationInfo.messages[activeMessageID].bodyMarkdown;

                            $currentSourceBtn.closest('.Thread__message').append(`
                                <div class="tb-source-field" id="tb-source-${activeMessageID}">
                                    <textarea readonly></textarea>
                                </div>`);
                            $(`#tb-source-${activeMessageID} textarea`).text(messageSource);
                        }
                    }
                });
            });
        }
```
